packages/op-coreutils/src/op_coreutils/bigquery/write.py

```diff
@@ -152,3 +161,55 @@ def _write_df_to_bq(df: pl.DataFrame, destination: str, job_config=bigquery.Load
         log.info(
             f"{operation_prefix}{job_config.write_disposition}: Wrote {human_rows(len(df))} {human_size(filesize)} to BQ {destination}"
         )
+
+
+def upsert_partition(
+    df: pl.DataFrame, dt: str, dataset: str, table_name: str, unique_keys=["dt"]
+):
+    """Upsert data into a specific partition in a BigQuery table.
+
+    Args:
+        df (pl.DataFrame): The DataFrame to upsert.
+        dt (str): The partition date in 'YYYY-MM-DD' format.
+        dataset (str): The BigQuery dataset name.
+        table_name (str): The BigQuery table name.
+        unique_keys (list, optional): Columns that uniquely identify rows. Defaults to ['dt'].
+    """
+    client = init_client()
+
+    df = df.with_columns(dt=pl.lit(dt).str.strptime(pl.Datetime, "%Y-%m-%d"))
+
+    destination = f"{dataset}.{table_name}"
+    staging_table_name = f"{table_name}_staging"
+    staging_destination = f"{dataset}.{staging_table_name}"
+
+    _write_df_to_bq(
+        df,
+        staging_destination,
+        job_config=bigquery.LoadJobConfig(
```

Comment:
For the staging table, we should use the WRITE_EMPTY write disposition. That way we make sure we are not colliding with something else possibly going on at the same time. 

For the same reason it might also be good to generate a small random suffix to use on the staging table name. That way the staging table will be guaranteed to be unique to this upsert call. 

We should also add an expiration time on the staging table, something like 30 minutes, just in case something breaks we know the table will be automatically deleted by BQ later on.